### PR TITLE
feat : Todo CRUD 및 Member 회원가입 및 인증, 인가 구현

### PR DESCRIPTION
**Pull request by jaeyeonme**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature

## What is the current behavior?

Issue Number: All

## PR information
1. Todo
- Todo 생성, 수정, 삭제, 조회 (CRUD)
- Todo BDD 테스트 코드 작성

2. Member
- 회원가입 (메일 전송) 및 메일 검증을 통해 회원가입
- Spring Security 를 통해 login 구현
- Member BDD 테스트 코드 작성

3. 리팩토링
- VO, DTO, Service, Mapper 등 코드 리팩토링

### CODE REVIEW

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/0404be960d468b47c16ad082ac42d5fef44b8a3e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/0404be960d468b47c16ad082ac42d5fef44b8a3e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/0404be960d468b47c16ad082ac42d5fef44b8a3e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/0404be960d468b47c16ad082ac42d5fef44b8a3e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/9bf63845234d668bbe4383fb034cd1902807b620/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/9bf63845234d668bbe4383fb034cd1902807b620/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/9bf63845234d668bbe4383fb034cd1902807b620/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/9bf63845234d668bbe4383fb034cd1902807b620/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/16b426806efc076cb4d68d470d4d8c1d27cfeef9/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/16b426806efc076cb4d68d470d4d8c1d27cfeef9/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/16b426806efc076cb4d68d470d4d8c1d27cfeef9/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/16b426806efc076cb4d68d470d4d8c1d27cfeef9/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/f99d860aab17aba165915b8d3eb86aa8879edc6e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/f99d860aab17aba165915b8d3eb86aa8879edc6e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/f99d860aab17aba165915b8d3eb86aa8879edc6e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/f99d860aab17aba165915b8d3eb86aa8879edc6e/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/com/flab/todo/member/MemberService.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+	private void validateDuplicateUser(SignUpRequest signUpRequest) {
+		boolean existsByEmail = memberMapper.existsByEmail(signUpRequest.getEmail());
+		if (existsByEmail) {
+			throw new IllegalArgumentException("이미 가입된 회원입니다.");
+		}
+	}
+
+	private void validatePassword(SignUpRequest signUpRequest) {
+		if (!signUpRequest.getPassword().equals(signUpRequest.getPasswordConfirm())) {
```
Comment (f-lab-kyucumber) [inline src/main/java/com/flab/todo/member/MemberService.java:73]: [디미터 법칙](https://prohannah.tistory.com/204)을 위반하고 있습니다. 이런 코드의 작성은 지양해주세요

DTO에 구현이 들어가는게 좀 이상할 수도 있는데 요런 단순한 부분이라면 메소드로 추출하시는게 보는 사람도 이해하기 쉽고 코드도 깔끔할 것 같아요.

**Review by f-lab-kyucumber — CHANGES_REQUESTED** [review]

File: src/main/java/com/flab/todo/common/config/mail/JavaMailService.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.Properties;
+
+import org.springframework.context.annotation.Configuration;
+import org.springframework.mail.javamail.JavaMailSenderImpl;
+
+@Configuration
+public class JavaMailService {
+
+	private final MailProperties mailProperties;
```
[the remark this answers is not in the document]
Comment (gimseonjin) [reply src/main/java/com/flab/todo/common/config/mail/JavaMailService.java:11]: 저 pr은 폐기했습니다. 저희가 코드 합치면서 문제가 조금 있어서 저 pr 폐기하고 저기에 남겨주신 피드백 일부를 가져와서 새로 만들었습니다. 다음부터는 조심하겠습니다.

File: src/main/java/com/flab/todo/common/config/security/PasswordEncoder.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.flab.todo.common.config.security;
+
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PasswordEncoder extends BCryptPasswordEncoder {
+}
```
[the remark this answers is not in the document]
Comment (gimseonjin) [reply src/main/java/com/flab/todo/common/config/security/PasswordEncoder.java:8]: 처음에는 그렇게 구현했는데 MemberService에서 PasswordEncoder를 사용하고 SecurityConfig에서 MemberService를 사용하니까 순환참조가 발생했습니다!

그래서 우선 별도의 Component로 분리했어요

File: src/main/java/com/flab/todo/common/config/security/SecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+			.httpBasic()
+			.and()
+			.sessionManagement()
+			.maximumSessions(1) // 설정한 최대 허용 세션의 수가 되었을 때 추가적인 인증 요청(세션 생성)이 있을 경우 어떻게 처리를 할지 정하는 api
+			.maxSessionsPreventsLogin(true)
+			.and()
+			.sessionCreationPolicy(SessionCreationPolicy.ALWAYS); // 항상 HttpSesion 을 만든다.
+
+		http.csrf()
+			.disable()
```
[the remark this answers is not in the document]
Comment (jaeyeonme) [reply src/main/java/com/flab/todo/common/config/security/SecurityConfig.java:59]: REST API 에서는 session 기반 인증과는 다르게 stateless 하기 때문에 서버에 인증 정보를 보관하지 않는다고 하여서 disable 하였습니다!

https://kchanguk.tistory.com/197?category=887999

File: src/main/java/com/flab/todo/member/MemberService.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+@RequiredArgsConstructor
+public class MemberService implements UserDetailsService {
+
+	private final PasswordEncoder passwordEncoder;
+	private final MemberMapper memberMapper;
+	private final JavaMailService javaMailService;
+
+	public void verifyEmailAndComplete(String token, String email) {
+		Member member = memberMapper.findByEmailAndEmailToken(email, token);
+		validateIsExistEmail(member);
```
[the remark this answers is not in the document]
Comment (gimseonjin) [reply src/main/java/com/flab/todo/member/MemberService.java:29]: valid => 값이 true false => 틀렸다 안 틀렸다
verify => 값을 확인 & 검증 => 너가 보낸 값이 이거 맞니???? => 확인

라고 구분해서 사용하려고 합니다!!

File: src/main/java/com/flab/todo/member/MemberService.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+			throw new IllegalArgumentException("Wrong Token");
+		}
+	}
+
+	private void completeSignUp(Member member) {
+		member.completeSignUp();
+		memberMapper.update(member);
+	}
+
+	public void sendSignUpEmail(SignUpRequest signUpRequest) {
```
[the remark this answers is not in the document]
Comment (gimseonjin) [reply src/main/java/com/flab/todo/member/MemberService.java:51]: 이 부분 네이밍은 수정해보겠습니다!

File: src/main/java/com/flab/todo/member/MemberService.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+		}
+	}
+
+	private void sendVerificationEmailWithToken(Member member) {
+		MailMessage mailMessage = MailMessage.makeVerifyMailFrom(member);
+		javaMailService.send(mailMessage);
+		log.info(
+			"Sent verification email to: " + member.getEmail() + " with link: " + MailMessage.makeEmailVerificationLink(
+				member.getEmailToken(), member.getEmail()));
+	}
```
[the remark this answers is not in the document]
Comment (jaeyeonme) [reply src/main/java/com/flab/todo/member/MemberService.java:84]: 넵 이부분은 분리해 보겠습니다.

File: src/main/java/com/flab/todo/common/exception/ErrorResponse.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+
+import lombok.Getter;
+
+@Getter
+public class ErrorResponse {
+	private int status;
+	private String message;
+	private String error;
```
[the remark this answers is not in the document]
Comment (gimseonjin) [reply src/main/java/com/flab/todo/common/exception/ErrorResponse.java:12]: 아 error는 범용적으로 많이 쓰인다고 해서 넣었습니다!

지금 로직에서는 딱히 필요없어서 제거하겠습니다!!

File: src/main/java/com/flab/todo/common/config/security/PasswordEncoder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gimseonjin) [reply src/main/java/com/flab/todo/common/config/security/PasswordEncoder.java:8]: 이 부분은 UserDetailService 구현체를 분리함으로써 내부 Bean으로 구현하도록 수정하겠습니다!

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/fae15269ece16b8f5a6e2bcdfc8dcdf409394aab/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/fae15269ece16b8f5a6e2bcdfc8dcdf409394aab/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/fae15269ece16b8f5a6e2bcdfc8dcdf409394aab/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/fae15269ece16b8f5a6e2bcdfc8dcdf409394aab/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/0d0fc9722753f70730970483a8fa2d7ebddc18a4/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/0d0fc9722753f70730970483a8fa2d7ebddc18a4/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/0d0fc9722753f70730970483a8fa2d7ebddc18a4/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/0d0fc9722753f70730970483a8fa2d7ebddc18a4/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/869e16550cdf7ad27b4ef4d1e90a73b4e8fbd659/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/869e16550cdf7ad27b4ef4d1e90a73b4e8fbd659/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/869e16550cdf7ad27b4ef4d1e90a73b4e8fbd659/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/869e16550cdf7ad27b4ef4d1e90a73b4e8fbd659/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/com/flab/todo/common/event/VerificationEmailEvent.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.context.ApplicationEvent;
+
+import com.flab.todo.database.entity.Member;
+
+import lombok.Getter;
+
+@Getter
+public class VerificationEmailEvent extends ApplicationEvent {
+
+	private final Member member;
```
Comment (gimseonjin) [inline src/main/java/com/flab/todo/common/event/VerificationEmailEvent.java:12]: @jaeyeonme 재연님!! 전 개인적으로 여기에 Member가 들어가는거보다 SimpleMailMessage가 들어가는 게 더 좋을 거 같은데 어떻게 생각하세요???

Member -> SimpleMailMessage 변환 로직은 VerificationEmailEvent 생성자에 넣구요!!

Comment (jaeyeonme) [reply src/main/java/com/flab/todo/common/event/VerificationEmailEvent.java:12]: @gimseonjin 오! 한번 그렇게 해볼까요?! 근데 그렇게 바꾸면 어떤점이 좋아지는지 잘 모르겠어요 ㅜ

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/0850fb89ed1c31c3585e17b20cff89afa4a19848/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/0850fb89ed1c31c3585e17b20cff89afa4a19848/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/0850fb89ed1c31c3585e17b20cff89afa4a19848/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/0850fb89ed1c31c3585e17b20cff89afa4a19848/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Review by f-lab-kyucumber — CHANGES_REQUESTED** [review]

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/42042297064bbe483534d796354c92740e522f6f/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/42042297064bbe483534d796354c92740e522f6f/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/42042297064bbe483534d796354c92740e522f6f/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/42042297064bbe483534d796354c92740e522f6f/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

File: src/test/java/com/flab/todo/member/MemberServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import com.flab.todo.common.dto.SignUpRequest;
+import com.flab.todo.database.entity.Member;
+
+class MemberServiceTest {
+	private PasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
+	private MemberMapper memberMapper = Mockito.mock(MemberMapper.class);
+	private ApplicationEventPublisher eventPublisher = Mockito.mock(ApplicationEventPublisher.class);
```
[the remark this answers is not in the document]
Comment (jaeyeonme) [reply src/test/java/com/flab/todo/member/MemberServiceTest.java:25]: 테스트 추가했습니다! 기존 이메일 성공 테스트 케이스에 `verify(eventPublisher, ...)` 만 넣어줬는데 
이렇게 하는게 맞을까요?!

**Comment by gimseonjin** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[TodoController.java](https://github.com/f-lab-edu/todo-list/blob/dc7cbe4c269833eeaf713a7e4883c53f3ee09636/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoController.java)|100%|:green_apple:|
|[TodoService.java](https://github.com/f-lab-edu/todo-list/blob/dc7cbe4c269833eeaf713a7e4883c53f3ee09636/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Ftodo%2FTodoService.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/f-lab-edu/todo-list/blob/dc7cbe4c269833eeaf713a7e4883c53f3ee09636/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberController.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/f-lab-edu/todo-list/blob/dc7cbe4c269833eeaf713a7e4883c53f3ee09636/src%2Fmain%2Fjava%2Fcom%2Fflab%2Ftodo%2Fmember%2FMemberService.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Review by f-lab-kyucumber — APPROVED** [review]

File: dockerfile [hunk 11]
```diff
@@ -0,0 +1,5 @@
+FROM adoptopenjdk/openjdk11
+WORKDIR /app
+COPY . ./
+RUN ./gradlew build
+ENTRYPOINT ["java","-jar","build/libs/todo-0.0.1-SNAPSHOT.jar"]
```
[the remark this answers is not in the document]
Comment (jaeyeonme) [reply dockerfile:5]: ENTRYPOINT 는 실행할 때 재정의할 수 없고, CMD는 실행할 때 재정의 가능한 걸로 알고 있습니다!
그래서 HEAP_OPTION 이나 설정파일들을 CMD 로 컨트롤 할 수 있을 것 같습니다!

**Pull request merged**